lightning/src/ln/functional_test_utils.rs

```diff
@@ -2664,8 +2671,13 @@ pub fn pass_claimed_payment_along_route<'a, 'b, 'c>(origin_node: &Node<'a, 'b, '
 							channel.context().config().forwarding_fee_base_msat
 						}
 					};
-					if $idx == 1 { fee += expected_extra_fees[i]; }
-					expect_payment_forwarded!(*$node, $next_node, $prev_node, Some(fee as u64), false, false);
+					let mut expected_extra_fee_limit = None;
+					if $idx == 1 {
+						fee += expected_extra_fees[i];
+						expected_extra_fee_limit = Some(expected_extra_fees[i] as u64);
```

Comment:
IIUC, `expected_extra_fees` should contain an exact fee, not a fee limit. Can we assert on the exact amount rather than `<=`?

Reply:
Hmm, fair enough. It might be easy-ish if you cherry-pick this commit from #2823: https://github.com/lightningdevkit/rust-lightning/pull/2823/commits/cbdb7ec54d6203f23373d369474a919ac51ec258 but I didn't look too closely. 

> FWIW, still debating whether they should though and this should be considered a bug?

IMO `mpp_retry_overpay` shouldn't be using the skimmed fees field that way, seems like overpaying to meet `min_htlc` should have its own separate field.

Reply:
Agree that the interception interface is probably fine as-is. 

Not sure I'm getting your point about the case where there is overpayment *and* skimmed fees going on at the same time? I wrote a small patch for asserting on an exact `PaymentForwarded::skimmed_fee` in tests and it seems fine: https://github.com/valentinewallace/rust-lightning/tree/2024-01-exact-skimmed-fees-check. Lmk what you think.

Reply:
> > Not sure I'm getting your point about the case where there is overpayment _and_ skimmed fees going on at the same time?
> 
> Sorry, I may not have communicated that properly above: the main question I wanted to raise is that it is a bit unclear to me what behavior we would expect / prefer when overpayment and fee-skimming is happening at the same time: Say we have `A -> B -> C`, with an expected payment size (before skimming) of 100sats reaching C. A is overpaying 20sats to meet HTLC min, i.e. 120sats are reaching B. If B is now skimming off 30 sats for the channel open to C, should we expect B to now forward 120-30=90 sats or 100-30=70sats to C after channel opening, i.e., get a total extra fee of 50sats or 30sats? I think we now implement the latter behavior, but technically it might be 'nicer' if an LSP would let any overpaid fees count towards the withheld fee it's already going to withhold? Probably doesn't matter all too much though, fine to leave it as is.

Ah, okay I see what you're saying. LSPs have the ability to do that now I believe, since they have the inbound amount and the expected outbound amount. So if they want to put the min-htlc-overpay fees towards their skimmed fee, they could just forward the expected amount. But it's not enforced, which I think is what you're suggesting. I agree the current API seems fine for now.